# CST1440 Assessment 1


<hr style="border:2px solid gray">

- Name: Emaan Fatima

- MISIS: M01041173

### Question 1: (Total: 18 marks)
In this question we will consider the Metropolitan Police
crime statistics. This data is available
at https://data.police.uk either as a batch download of .csv files or through the (well
documented) police API. 

**a) (2 marks)**

Load the **Metropolitan Police** crime data for the period **August 2023 - July 2025** into a
single pandas dataframe.

 - The optional outcomes and stop and search data is not needed.

 Write code to remove all variables apart from **Month, Location, LSOA name, and Crime
Type.**

In [17]:
'''
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
 - Question 1(a) Load Metropolitan Police crime data (August 2023 - July 2025)
 - Columns retained: Month, Location, LSOA name, Crime type
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
'''
# Import required modules for file handling and data analysis
import zipfile
import os
import pandas as pd

# Extract all files from the ZIP folder to current repository 
with zipfile.ZipFile("Police.zip", "r") as zip_file:
    zip_file.extractall(".")  

# Verify extracted files
print(os.listdir())          

# Initialize a list to store DataFrames
crime_dfs = []

# Collect all folders and subfolders in a list
for subdir, dirs, files in os.walk("."):
    for file in files:
        if file.endswith(".csv") and "metropolitan-street" in file.lower():
            csv_path = os.path.join(subdir, file)
            crime_dfs.append(pd.read_csv(csv_path))  # Read each CSV

# Combine CSVs into one DataFrame
df = pd.concat(crime_dfs, ignore_index=True)

# Keep only the relevant columns for analysis
columns_to_keep = ["Month", "Location", "LSOA name", "Crime type"]
df = df[columns_to_keep]

# Check the loaded data
print(df.head())
print("\nUnique months loaded:", df["Month"].unique())
print("\nTotal rows:", len(df))

['.bash_history', '.cursor', '.git', '.gitconfig', '.idlerc', '.ipynb_checkpoints', '.ipython', '.junie', '.jupyter', '.lesshst', '.matplotlib', '.ssh', '.vscode', '2023-08', '2023-09', '2023-10', '2023-11', '2023-12', '2024-01', '2024-02', '2024-03', '2024-04', '2024-05', '2024-06', '2024-07', '2024-08', '2024-09', '2024-10', '2024-11', '2024-12', '2025-01', '2025-02', '2025-03', '2025-04', '2025-05', '2025-06', '2025-07', '2025-08', '3D Objects', '77t.zip', 'AL LABBB 5.ipynb', 'AppData', 'Application Data', 'Contacts', 'Cookies', 'CST1440 Assessment 1.ipynb', 'CST1440 Week 2.ipynb', 'DLLs', 'Doc', 'Documents', 'Downloads', 'etc', 'Favorites', 'filename.png', 'include', 'IntelGraphicsProfiles', 'java_error_in_pycharm.hprof', 'lab1.ipynb', 'Lib', 'libs', 'LICENSE.txt', 'Links', 'Local Settings', 'Music', 'My Documents', 'MYSUBMISSION', 'NetHood', 'new.txt', 'NEWS.txt', 'NTUSER.DAT', 'ntuser.dat.LOG1', 'ntuser.dat.LOG2', 'NTUSER.DAT{2ad838bc-efea-11ee-a54d-000d3a94eaa1}.TM.blf', 'NTUSER

**b) (3 marks)**

 - What are the different types of crime that appear in this dataset?

 - Which London Borough (“LSOA name”) had the most incidents of Anti-social behaviour?

 - Which London Borough (“LSOA name”) had the most incidents of Bicycle theft in June
2024?

In [18]:
'''
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
 - Question 1(b): Crime Analysis
 - List unique crime types and dispaly 'LSOA name' with highest incidents
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
'''

# Decorative Header
print("=" * 100)
print("Crime Analysis".center(100))
print("=" * 100)

# (1️) List unique crime types from dataset
print("● Crime Types:")
for crime_type in df['Crime type'].unique(): # iterate and print each unique crime type
    print(f' - {crime_type}')
print("-" * 85)

# (2) Find LSOA with the Most Anti-social behaviour incidents
top_anti_social_lsoa  = df.loc[df['Crime type'] == 'Anti-social behaviour', 'LSOA name'].value_counts().head(1) # Count incidents and keep top LSOA
print("● Highest Anti-social behaviour incidents:")
print(f' - {top_anti_social_lsoa }')
print("-" * 85)

# (3) Find LSOA with Most Bicycle theft incidents (June 2024)
top_bike_theft= df.loc[
    (df["Crime type"] == "Bicycle theft") & (df["Month"] == "2024-06"), 
    "LSOA name"
    ].value_counts().head(1) # Select row by crime type and month
print("● Highest Bicycle theft incidents (June 2024):")
print(f' - {top_bike_theft}')

# Decorative Footer
print("=" * 100)
print("Program End".center(100))
print("=" * 100)


                                           Crime Analysis                                           
● Crime Types:
 - Violence and sexual offences
 - Drugs
 - Anti-social behaviour
 - Burglary
 - Criminal damage and arson
 - Other theft
 - Possession of weapons
 - Public order
 - Vehicle crime
 - Robbery
 - Shoplifting
 - Other crime
 - Bicycle theft
 - Theft from the person
-------------------------------------------------------------------------------------
● Highest Anti-social behaviour incidents:
 - LSOA name
Hillingdon 031A    10247
Name: count, dtype: int64
-------------------------------------------------------------------------------------
● Highest Bicycle theft incidents (June 2024):
 - LSOA name
Kingston upon Thames 009C    60
Name: count, dtype: int64
                                            Program End                                             


**c) (9 marks)**

 - Produce at least two graphs to illustrate the data. Interpret your graphs and write two
sentences explaining what your graphs show.

 - Choose some variables to focus on and calculate some appropriate summary statistics. You
should demonstrate your understanding of measures of centrality, spread, and quantiles.

 - Interpret your statistics, and write four sentences explaining to a layperson what your results
say about crime in London.

In [19]:
'''
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
 - Question 1(c): Graph 1 - Bar Chart 
 - Display top 10 areas (LSOA names) with the highest number of crimes.
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
'''
# --- Setup Lets-Plot for plotting
from lets_plot import *
LetsPlot.setup_html()

# --- Decorative Header 
print("=" * 100)
print("Crime Analysis – Bar Chart".center(100))
print("=" * 100)

# --- Find top 10 LSOA areas with the highest crime rate
crime_rate = df['LSOA name'].value_counts().head(10)

# --- Convert Series to DataFrame and rename columns 
# Converting to DataFrame for easier column naming and plotting
crime_rate_df = crime_rate.reset_index()
crime_rate_df.columns = ['LSOA name', 'Crime Rate']

# --- Create bar chart to visualize crime rates by area
graph_1 = ggplot(crime_rate_df, aes(x='LSOA name', y='Crime Rate')) + \
    geom_bar(stat='identity', fill = '#3CB371', color='black', width=0.6) + \
    theme(axis_text_x=element_text(angle=50, size=13, color='#301934')) + \
    ggtitle("Top 10 Areas with Most Crimes") + \
    xlab("Area Name (LSOA)") + ylab("Crime Rate")

# --- Display the chart 
graph_1


                                     Crime Analysis – Bar Chart                                     


In [20]:
# Decorative Header
print('=' * 40)
print(" - Graph 1 Interpretation:")
print('=' * 40)

# ------- Graph 1 - Interpretation -------- #
print("This bar chart shows that most London neighborhoods have relatively low crime rates, while a few areas")
print("such as Westminster 013G, 013B, and 018A experience extremely high crime rates, exceeding 90,000 crimes,")
print("several times higher than the lower-ranked neighborhoods.")

 - Graph 1 Interpretation:
This bar chart shows that most London neighborhoods have relatively low crime rates, while a few areas
such as Westminster 013G, 013B, and 018A experience extremely high crime rates, exceeding 90,000 crimes,
several times higher than the lower-ranked neighborhoods.


In [21]:
'''
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
- Question 1(c): Graph 2 (Line + Scatter Plot)
- Dispaly monthly crime trends in London. (August 2023 - August 2025)
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
'''

# --- Decorative Header 
print("=" * 100)
print("Crime Analysis – (Line + Scatter Plot)".center(100))
print("=" * 100)

# --- Find total crimes rate for each month 
#  Easier to name columns and plot by converting series to a DataFrame.
monthly_incidents = df['Month'].value_counts().sort_index().reset_index()
monthly_incidents.columns = ['Month', 'Crime Rate']

# --- Create line + scatter plot 
# Line shows trend over time; points highlight individual months
graph_2 = ggplot(monthly_incidents, aes(x='Month', y='Crime Rate')) + \
    geom_line(color='blue', size=2.5) + \
    geom_point(color='red', size=3.5) + \
    ggtitle("Monthly Crime Trend in London (August 2023 - August 2025)") + \
    xlab("Month of the Year") + ylab("Total Crimes") + \
    theme(axis_text_x=element_text(angle=50, size=13))

# --- Display the chart 
# Visualize monthly crime trends to identify peaks and drops
graph_2


                               Crime Analysis – (Line + Scatter Plot)                               


In [22]:
# Decorative Header
print('=' * 40)
print(" - Graph 2 Interpretation:")
print('=' * 40)

# ------- Graph 2 Interpretation --------
print("This line and scatter plot shows the monthly trend of total crimes in London from August 2023 to")
print("August 2025. Crime rates fluctuate each month, with peaks and drops clearly visible. July 2025 had")
print("the highest number of crimes, exceeding 500,000, while February 2025 recorded the lowest, around 334,000.")


 - Graph 2 Interpretation:
This line and scatter plot shows the monthly trend of total crimes in London from August 2023 to
August 2025. Crime rates fluctuate each month, with peaks and drops clearly visible. July 2025 had
the highest number of crimes, exceeding 500,000, while February 2025 recorded the lowest, around 334,000.


- Choose some variables to focus on and calculate some appropriate summary statistics. You should demonstrate your understanding of measures of centrality, spread, and quantiles.

- Interpret your statistics, and write four sentences explaining to a layperson what your results say about crime in London.



In [23]:
'''
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
 - Question 1(c):   Statistics summary of selected variables
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
'''

# --- Decorative Header 
print("=" * 100)
print("Statistical Summary".center(100))
print("=" * 100)


# --- (1️) Crimes per Month
# Count crimes for each month and sort chronologically
monthly_crimes = df['Month'].value_counts().sort_index()

# Compute statistics for monthly crime rates
mean = monthly_crimes.mean()
median = monthly_crimes.median() 
std = monthly_crimes.std() # Variation across months
variance = monthly_crimes.var()
quantiles = monthly_crimes.quantile([0.25, 0.5, 0.75]) # Quartiles to show distribution

# Display results
print(" ===== Crimes per Month =====")
print(f"● Mean: {mean}")
print(f"● Median: {median}")
print(f"● Std: {std}")
print(f"● Variance: {variance}")
print("● Quantiles:\n", quantiles)
print()


# --- (2️) Crimes per LSOA / Neighborhood
# Count crimes per neighborhood to find high and low crime areas
crime_areas = df['LSOA name'].value_counts()

# Compute statistics for every LSOA
mean  = crime_areas.mean()
median = crime_areas.median()
std  = crime_areas.std()
quantiles = crime_areas.quantile([0.10, 0.50, 0.90]) # 10th, 50th (median), and 90th percentiles

# Display results
print(" ===== Crimes per LSOA =====")
print(f"● Mean: {mean}")
print(f"● Median: {median}")
print(f"● Std: {std}")
print("● Quantiles:\n", quantiles)
print()


# --- (3️) Crimes per Crime Type
# Count total occurrences for each crime type
crime_types = df['Crime type'].value_counts() 

# Compute statistics for crime types
mean = crime_types.mean()
median= crime_types.median()
std = crime_types.std()
crime_min = crime_types.min()
crime_max = crime_types.max()
crime_range = crime_max - crime_min
quantiles = crime_types.quantile([0.25, 0.5, 0.75])  # 25th, 50th (median), and 75th percentiles for crime types

# Display results 
print(" ===== Crimes per Crime Type =====")
print(f"● Mean: {mean}")
print(f"● Median: {median}")
print(f"● Std: {std}")
print(f"● Min: {crime_min}")
print(f"● Max: {crime_max}") 
print(f"● Range: {crime_range}")
print("● Quantiles:\n", quantiles)

                                        Statistical Summary                                         
 ===== Crimes per Month =====
● Mean: 397573.0
● Median: 387212.0
● Std: 48379.53983951618
● Variance: 2340579875.0833335
● Quantiles:
 0.25    366364.0
0.50    387212.0
0.75    400964.0
Name: count, dtype: float64

 ===== Crimes per LSOA =====
● Mean: 1019.7912325941206
● Median: 375.0
● Std: 2755.9154288147747
● Quantiles:
 0.1       4.0
0.5     375.0
0.9    2450.0
Name: count, dtype: float64

 ===== Crimes per Crime Type =====
● Mean: 709951.7857142857
● Median: 488548.5
● Std: 674594.9582334777
● Min: 42023
● Max: 2275281
● Range: 2233258
● Quantiles:
 0.25    305531.5
0.50    488548.5
0.75    814841.5
Name: count, dtype: float64


=============
Interpretation for Layperson
=============

**● Monthly crime:**
Each month, London records about **398,000 crimes on average**. The **month in the middle** has roughly **387,000 crimes,**
but numbers can vary by around **48,000**, showing crime rates fluctuate throughout the year. 
Talking about the **distribution**, **25% of months** have fewer than **366,000 crimes,** **half** have up to **387,000** and 
the **busiest 25%** exceed **401,000 crimes.** Overall, throughout the year, crime rate fluctuates.

**● Neighborhood crime:**
Albeit most neighborhoods see **relatively few crimes**, with a **median** of **375** but the average is higher **(≈1,020)** due
to some high-crime areas. Crime varies widely, with 10% of neighborhoods having **4 
crimes** and the top 10% having **2,450 crimes**. While most areas are fairly safe, a few areas 
experience much higher crime levels.

**● Crime type:**
Most crime types occur at moderate levels, **with a median of 489,000**, but the average is higher
**(≈710,000)** because a few crime types are very common. The least frequent type has about
**42,000 incidents** and the most common being **roughly 2.2 million incidents**. This shows a few crime types 
dominate London’s overall crime figures, while most are relatively rare.

=============
=============

**f) (4 marks)**

 - Choose a “Location” with between 200 and 500 total reported crimes.

 - For this location calculate a 95% **confidence interval** for the **proportion** of “Other theft”
crimes.

 - Interpret your output and write two sentences explaining to an expert how you used probability
theory in your calculations.

In [24]:
# --------- Filter locations which fall under 200–500 total crimes

loc_count = df['Location'].value_counts()  # Count total crimes per location
available_loc = loc_count[(loc_count >= 200) & (loc_count <= 500)] # Keep only locations within 200–500 crimes

# Display the filtered locations
print("Available Locations:")
print(available_loc)

Available Locations:
Location
On or near Mount Mills             500
On or near Stanton Road            500
On or near Mount Adon Park         500
On or near Gunnersbury Crescent    500
On or near Greenwood Close         500
                                  ... 
On or near Shirley Grove           200
On or near St James'S Square       200
On or near Linden Way              200
On or near Masefield Crescent      200
On or near Castlewood Road         200
Name: count, Length: 7939, dtype: int64


In [25]:
'''
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
Question 1(f): 95% CI for Proportion of 'Other theft' Crimes
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
'''

# Import Numpy for calculations
import numpy as np

# Selected location for Confidence Interval
location = "On or near Shirley Grove"

# Filter crimes at the chosen location for 'Other theft'
crimes = (df.loc[df['Location'] == location, 'Crime type'] == "Other theft")

# Compute sample proportion and count of 'Other theft'
p = crimes.mean()
n = crimes.count()

# Standard error for sample proportion
SE = np.sqrt(p * (1 - p) / n)

# Z-Value for Confidence Level
z = 1.96

# 95% Confidence Interval - lower and upper bounds of the proportion
lower = p - z * SE
upper = p + z * SE

# Decorative Header
print("=" * 100)
print("Confidence Interval".center(100))
print("=" * 100)

# Display results
print(f" - Sample Proportion (p) : {p:.2f}")
print(f" - 95% Confidence Interval: ({lower:.2f}, {upper:.2f})")

# Section Header
print("\n" + "=" * 50)
print("\tConfidence Interval Interpretation:")
print("=" * 50)

# --- Interpretation for layperson ---
print(" ● Using the 95% confidence interval, we’re confident that the true proportion of 'Other theft' ")
print("   crimes at 'On or near Shirley Grove' is between 2% and 7%.\n")

print(" ● We calculated the sample proportion and used probability theory to find the standard error. ")
print("   Then, using the normal distribution, we constructed the 95% confidence interval, which is 2% to 7%.")

                                        Confidence Interval                                         
 - Sample Proportion (p) : 0.04
 - 95% Confidence Interval: (0.02, 0.07)

	Confidence Interval Interpretation:
 ● Using the 95% confidence interval, we’re confident that the true proportion of 'Other theft' 
   crimes at 'On or near Shirley Grove' is between 2% and 7%.

 ● We calculated the sample proportion and used probability theory to find the standard error. 
   Then, using the normal distribution, we constructed the 95% confidence interval, which is 2% to 7%.


## Question 2 (Total: 14 marks)

Consider the following Python code, that represents some of the knowledge in this module so
far.

```python 
knowledge = [
("Alan Turing", "isa", "Person"),
("Alan Turing", "field", "Computer Science"),
("Computer Science", "subfieldOf", "STEM"),
("Machine Learning", "subfieldOf", "Artificial Intelligence"),
("Artificial Intelligence", "subfieldOf", "Computer Science"),
("STEM", "subfieldOf", "Academia"),
("LLM", "subfieldOf", "Artificial Intelligence")
]


**a) (5 marks)**
  
Provide a short user-guide (perhaps using the print command) explaining

• what kind of Knowledge Representation is being used here

• the features/limitations of this representation,

• how this representation could be searched.


In [26]:
'''
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
Question 2(a): Knowledge Representation Guide
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
'''

# Decorative Header
print("="*80)
print("💡 Knowledge Representation".center(80))
print("="*80)

# (1) Knowledge Representation Type                                 
print("\n📌 Knowledge Representation Type:")
print('- Semantic Network is being used here for knowledge representation.')
print("This is a Semantic Network because it represents concepts as nodes and their relationships as links between them.")
print("For instance: ('Mathematics', 'subfieldOf', 'STEM') shows a concept connected to another.")

# (2) Features and Limitations  
print("\n📌 Features / Limitations:")

print("- Features:")
print("  • Easy to visualize: Shows how concepts are connected clearly.")
print("  • Flexible: New concepts or relationships can be added easily.")
print("  • Supports discovery: Related information can be found by following links.\n")

print("- Limitations:")
print("  • Challenging reasoning: Some inferences require following several connected concepts.")
print("  • Indirect connections: Some concepts may not be directly linked. ")

# (3) Search Methods
print("\n📌 Search Methods:")
print("• Forward Chaining:")
print("  In a semantic network, Forward chaining can be used to discover new connections by following links.\n")

print("• Depth-First Search (DFS):")
print("  DFS can also be used which works by discovering a path of connected concepts as far as possible before backtracking.\n")

print("• Pattern Matching:")
print("  We can also search the knowledge base for specific concepts or relationships using pattern matching.")
print('  For instance: query(knowledge, subject = "Alan Turing")')

# Decorative Footer
print("="*80)
print("End of Guide".center(80))
print("="*80)


                           💡 Knowledge Representation                           

📌 Knowledge Representation Type:
- Semantic Network is being used here for knowledge representation.
This is a Semantic Network because it represents concepts as nodes and their relationships as links between them.
For instance: ('Mathematics', 'subfieldOf', 'STEM') shows a concept connected to another.

📌 Features / Limitations:
- Features:
  • Easy to visualize: Shows how concepts are connected clearly.
  • Flexible: New concepts or relationships can be added easily.
  • Supports discovery: Related information can be found by following links.

- Limitations:
  • Challenging reasoning: Some inferences require following several connected concepts.
  • Indirect connections: Some concepts may not be directly linked. 

📌 Search Methods:
• Forward Chaining:
  In a semantic network, Forward chaining can be used to discover new connections by following links.

• Depth-First Search (DFS):
  DFS can also be used 

**b) (4 marks)**

Add ten more triples to represent more of the knowledge from this module. You should try to
acheive a greater breadth and depth in this knowledge representation

In [27]:
'''
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
Question 2(b): Appending Knowledge Representation 
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
'''

# Existing Knowledge Representation
knowledge = [ ("Alan Turing", "isa", "Person"), 
              ("Alan Turing", "field", "Computer Science"),
              ("Computer Science", "subfieldOf", "STEM"),
              ("Machine Learning", "subfieldOf", "Artificial Intelligence"), 
              ("Artificial Intelligence", "subfieldOf", "Computer Science"), 
              ("STEM", "subfieldOf", "Academia"),
              ("LLM", "subfieldOf", "Artificial Intelligence") ]

# New knowledge to append 
new_knowledge = [
    ("Python", "usedFor", "Artificial Intelligence"),
    ("Python", "usedFor", "Data Science"),
    ("Data Science", "subfieldOf", "Mathematics"),
    ("Mathematics", "subfieldOf", "STEM"),
    ("Statistics", "subfieldOf", "Data Science"),
    ("Mean", "partOf", "Statistics"),
    ("Standard Deviation", "partOf", "Statistics"),
    ("Probability", "subfieldOf", "Statistics"),
    ("Random Variables", "partOf", "Probability"),
    ("Bayes Theorem", "partOf", "Probability")
]

# Append 10 more triples to existing knowledge
knowledge += new_knowledge

# Decorative Header
print('=' * 55)
print(f'New Knowledge Representation:'.center(50))
print('=' * 55)

# Display the updated semantic network
for subjects, relations, objects in knowledge:
    print(f"({subjects} , {relations}  , {objects})")

          New Knowledge Representation:           
(Alan Turing , isa  , Person)
(Alan Turing , field  , Computer Science)
(Computer Science , subfieldOf  , STEM)
(Machine Learning , subfieldOf  , Artificial Intelligence)
(Artificial Intelligence , subfieldOf  , Computer Science)
(STEM , subfieldOf  , Academia)
(LLM , subfieldOf  , Artificial Intelligence)
(Python , usedFor  , Artificial Intelligence)
(Python , usedFor  , Data Science)
(Data Science , subfieldOf  , Mathematics)
(Mathematics , subfieldOf  , STEM)
(Statistics , subfieldOf  , Data Science)
(Mean , partOf  , Statistics)
(Standard Deviation , partOf  , Statistics)
(Probability , subfieldOf  , Statistics)
(Random Variables , partOf  , Probability)
(Bayes Theorem , partOf  , Probability)


**c) (5 marks)**

**Write a function `query` that returns all triples matching a given pattern.**  
It should be possible to specify a pattern for each variable in the triple.

**For example:**  

```python
query(knowledge, subject = "Alan Turing")


should return:

```python
[
    ('Alan Turing', 'isa', 'Person'),
    ('Alan Turing', 'field', 'Computer Science'),
    # along with any other appropriate triples you’ve added
]

And:
```python
query(knowledge, object = "Artificial Intelligence")

should return:

```python
[
    ('Machine Learning', 'subfieldOf', 'Artificial Intelligence'),
    ('LLM', 'subfieldOf', 'Artificial Intelligence'),
]


Along with any other appropriate triples you’ve added

In [28]:
'''
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
Question 2(c): Query Function for Knowledge Triples
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
'''

# Define Query Function for Knowledge Triples
def query(knowledge, subject=None, relation=None, object=None):
    
    # Iterate through all triples in the knowledge base
    for triple in knowledge:
        # Skip triples that don't match the specified conditions
        if subject != None and triple[0] != subject:
            continue
        if relation != None and triple[1] != relation:
            continue
        if object != None and triple[2] != object:
            continue
        print(triple)  # Display triple that meets the requirement

# Decorative header
print('=' * 100)
print(f'Query Function for Knowledge Triples'.center(100))
print('=' * 100)

# Demonstrate querying the knowledge base
print("1. Subject = 'Alan Turing'")
query(knowledge, subject="Alan Turing")
print()

print("2. Object = 'Artificial Intelligence'")
query(knowledge, object="Artificial Intelligence")
print()

print("3. Relation = 'isa'")
query(knowledge, relation="isa")
print('=' * 100)


                                Query Function for Knowledge Triples                                
1. Subject = 'Alan Turing'
('Alan Turing', 'isa', 'Person')
('Alan Turing', 'field', 'Computer Science')

2. Object = 'Artificial Intelligence'
('Machine Learning', 'subfieldOf', 'Artificial Intelligence')
('LLM', 'subfieldOf', 'Artificial Intelligence')
('Python', 'usedFor', 'Artificial Intelligence')

3. Relation = 'isa'
('Alan Turing', 'isa', 'Person')


## Question 3: (Total: 8 marks)
**(a) (5 marks)**

**Consider a perceptron (a single artificial neuron) that consists of 5 input nodes with weights**  
w₁ = 0.2, w₂ = 0.4, w₃ = 0.3, w₄ = 0.6, w₅ = 0.2  

**and an activation function defined by:**  

\[
f(x) = 
\begin{cases} 
0 & \text{if } x < 1 \\
1 & \text{if } x \ge 1
\end{cases}\]

**Write a function `perceive(x)` that takes a numpy array of length 5 and returns the output of
the perceptron.**

---
For example
```python
perceive([0.2, 0.3, 0.1, 0.5, 0.2])




should return 0, and
```python
perceive([0.5, 0.5, 0.1, 1, 1])

should return 1.


In [29]:
'''
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
Question 3(a): Perceptron Function
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
'''

# Import Numpy for calculations
import numpy as np

# Define function that computes output from inputs and weights
def perceive(x):
    # Store the weights for each input in a NumPy array
    weights = np.array([0.2, 0.4, 0.3, 0.6, 0.2])
    
    # Compute the combined input value by multiplying the weights
    total_input = np.sum(weights * x)
    
    # Output 1 if perceptron activates, else 0
    return 1 if total_input >= 1 else 0

# Decorative Header
print('=' * 55)
print(f'Perceptron Function'.center(50))
print('=' * 55)

# Test inputs for the perceptron
test_inputs = [
    np.array([0.2, 0.3, 0.1, 0.5, 0.2]),
    np.array([0.5, 0.5, 0.1, 1, 1]),
    np.array([0.9, 0.4, 0.3, 0.8, 0.5])
]

# Display perceptron outputs for each test input
for x in test_inputs:
    print(f"Input: {x}  → Output: {perceive(x)}")

# Decorative Footer
print('=' * 60)
print('Function End'.center(50))
print('=' * 60)

               Perceptron Function                
Input: [0.2 0.3 0.1 0.5 0.2]  → Output: 0
Input: [0.5 0.5 0.1 1.  1. ]  → Output: 1
Input: [0.9 0.4 0.3 0.8 0.5]  → Output: 1
                   Function End                   


**(b) (3 marks)**

**Task:** Provide a short user-guide (perhaps using the print command) explaining  

1. **Where knowledge is being represented in this model**  
2. **How this knowledge is acquired for this type of Knowledge Representation**

---

In [30]:
'''
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
Question 3(b): Knowledge Representation Guide
∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞∞
'''

# Decorative Header
print('=' * 70)
print("Knowledge Representation Guide".center(70))
print('=' * 70)

# Knowledge representation explanation
print("1. Knowledge Representation")
print("  Here knowledge is being represented in the form of weights, e.g., [0.2, 0.3, 0.1, 0.5, 0.2].")
print("  These weights store how each input influences the output, which is a role of a knowledge base.")

# How knowledge is acquired
print("\n2. How is this knowledge acquired?")
print("   In this case, the knowledge is acquired from predefined weights.")
print("   The weights are manually assigned by the programmer, so the perceptron uses them as-is.")
print("=" * 70)

                    Knowledge Representation Guide                    
1. Knowledge Representation
  Here knowledge is being represented in the form of weights, e.g., [0.2, 0.3, 0.1, 0.5, 0.2].
  These weights store how each input influences the output, which is a role of a knowledge base.

2. How is this knowledge acquired?
   In this case, the knowledge is acquired from predefined weights.
   The weights are manually assigned by the programmer, so the perceptron uses them as-is.


#                                                                        ≿━━━━༺❀༻━━━━≾

 ## Bonus mark - PEP0008

**# PEP 8 Breach 1 – Formatting issue**

```python
previous_balance = 1379
transaction = 200
current_balance = previous_balance - transaction
print(f'Your previous balance was £{previous_balance} \
you have withdrawn £{transaction} \
so your new balance is £{current_balance}')

In [75]:
# === Problem ===
# The issue in this code is the use of backslash (\) line continuation inside
# an f-string, which is not recommended by PEP 8. We should use implicit
# line continuation (via parentheses) instead.

# According to PEP 8:
# "The preferred way of wrapping long lines is by using Python’s implied
# line continuation inside parentheses, brackets, and braces. These should
# be used in preference to using a backslash for line continuation."

# This code therefore breaks one guideline:
# ❌ Uses backslashes (\) instead of parentheses for continuation

# --- This issue was identified from week 2 HTML document ---#


**PEP 8 Format Correction 1**

```python
previous_balance = 1379
transaction = 200
current_balance = previous_balance - transaction
print(
    f"Your previous balance was £{previous_balance}, "
    f"you have withdrawn £{transaction}, "
    f"so your new balance is £{current_balance}."
)



In [72]:
# === Solution ===
# A better way to write this is by using parentheses for implicit line
# continuation.

**PEP 8 Breach 2 – Formatting issue**

```python
df.sort_values(by = 'Enrolled')

In [76]:
# === Problem ===
# The issue in this code is the use of extraneous whitespace around the 
# assignment operator inside a function call. PEP 8 recommends **no spaces**
# around the '=' sign for keyword arguments.

# According to PEP 8:
# "Don’t use spaces around the '=' sign when used to indicate a keyword 
# argument, or when used to indicate a default value for an unannotated 
# function parameter."

# This code therefore breaks one guideline:
# ❌ Uses spaces around '=' in a keyword argument

# --- This issue was identified from week 2 HTML document ---#

**PEP 8 Format Correction 1**

```python
df.sort_values(by='Enrolled')


In [77]:
# === Solution ===
# A better way to write this is without spaces around the '=' sign in the
# keyword argument


## ============================ Thank You! =========================